docs/rholang/rholangmatchingtut.md

```diff
@@ -0,0 +1,201 @@
+
+    for( x <- @{z!(Nil)} ){ Nil }
+
+won't compile, due to the globally free variable `z`,
+
+    for( _ <- @{Nil} ){ {Nil} \/ {@Nil!(Nil)} }
+
+is incorrect, due to a logical connective being where a process should be, and
+
+    for( x <- @{[1, 2 ... z]}){ Nil }
+
+won't compile because `z` is free and the given list is a pattern.
+
+## Name Equivalence
+
+The RHO calculus, on which Rholang is based, says that two names are equivalent when the processes that they quote are equivalent, and that processes are equated via the relations below (and nothing more). Here, `P`, `Q` and `R` are processes:
+
+* `P | Q = Q | P` (commutativity)
+* `P = P | Nil = Nil | P` (identity)
+* `(P | Q) | R = P | (Q | R)` (associativity)
+
+Therefore, `@ P | Q = @ Q | P` and `@ P | Nil = @ P = @ Nil | P`, etc.
+
+In Rholang, these relations only apply *to the top level* of any name. In addition to the three given above, we also evaluate expressions on the top level. So for example, as channels,
+
+* `@{10 + 2} = @{5 + 7}` and
+* `@{for(x <- @Nil){ 10 + 2 }} = @{for(x <- @Nil){ 5 + 7 }}`.
+
+Since we use variables, channels also respect alpha equivalence, meaning that, for example, as channels
+
+* `@{for( x <- @Nil ){ Nil }}  = @{for( z <- @Nil ){ Nil }}`.
+
+In the RHO calculus we don't have to worry about distinguishing the top level from other parts of a channel, but because of things like pattern-matching, we have to in Rholang. This will be relevant later on, where there are restrictions on pattern-matching because of this.
+
+## "Looking through the looking glass": Patterns Within Patterns
+
+Patterns follow the same rules as channels in terms of equivalence. In the case of -arity matching, joins, logical connectives, etc. the rules apply to each name pattern individually.
+
+Here we need to treat statements that are not on the top level. We reemphasize that the rules for name equivalences described in the last section only apply on the top level. When we are matching *patterns within patterns*, these equivalence rules do not apply and we check for an exact text match, **excluding** alpha equivalence. This means that the following two channels are *not* equivalent:
+
+
+* `@for(  @for( @{ Nil | x } <- @Nil ){ Nil } <- @Nil ){ Nil }}`
+* `@for(  @for( @{ x | Nil } <- @Nil ){ Nil } <- @Nil ){ Nil }}`.
+
+Note that while the patterns `@{ x | Nil }}` and `@{ Nil | x }` are equivalent at the top level, the patterns
+
+* `@for( @{ Nil | x } <- @Nil ){ z }`
+* `@for( @{ x | Nil } <- @Nil ){ z }`
+
+are *not*. When we are matching *patterns within patterns*, these equivalence rules do not apply and we check for an exact text match, up to alpha equivalence. The principle is that you can only look through the looking glass once: when checking name equivalence or matching a pattern to a name or a process, if a given pattern is part of a larger pattern, there has to be an exact match.
```

Comment:
This isn't true. Those both reduce down to `@{ x }`

Reply:
@pyrocto re our conversations I was under the impression that patterns within patterns were exact text matches?

Reply:
"Text match" is approximately accurate. We normalize the process, and then require an exact match.

Reply:
This clears things up. I'll update this soon.